056-zk-kraft-migration.md

```diff
@@ -0,0 +1,247 @@
+
+> NOTE: this kind of approach is already in place for the Kafka version upgrade.
+
+Assume the following prerequisites, with a ZooKeeper based Kafka cluster defined by:
+
+* A `Kafka` custom resource.
+* One or more `KafkaNodePool`(s) for running the ZooKeeper based brokers.
+
+It would assume the operator is not using KRaft by default nor has the `+UseKRaft` feature gate enabled.
+
+First step for the user would be enabling the `+UseKRaft` feature gate unless it’s already enabled by default in a new operator release (depending on which Strimzi version this proposal will land).
+So it could be needed doing an upgrade of the current operator to the new release.
+In both cases, the operator is rolled, a new one starts and it’s able to handle the creation of new Apache Kafka clusters in KRaft mode but not migrating the old ZooKeeper based ones automatically.
+
+Currently, the operator prints the following warning:
+
+```shell
+io.strimzi.operator.common.model.InvalidResourceException: Tha Kafka cluster my-cluster is invalid: [At least one KafkaNodePool with the controller role and at least one replica is required when KRaft mode is enabled]
+```
+
+This is because with KRaft mode enabled in the operator, the reconciled `Kafka` resource doesn’t have any controller nodes (in any pool).
+
+The migration process can be also rolled back to ZooKeeper at any point before taking the KRaft controllers out of the migration mode.
+
+### Finite State Machine (FSM)
+
+The migration process can be described with a Finite State Machine (FSM).
+The operator has a new `MigrationManager` component in charge of tracking the migration finite state machine.
+Transitioning across the states depends on the current state, stored in the `Kafka.status.migrationState` field, and the "input" represented by the user interaction with the operator through the following annotations:
+
+* `strimzi.io/kraft-migration`: to be applied on the `Kafka` custom resource with `true|false` values to start/end the overall migration process.
+* `strimzi.io/kraft-migration`: to be applied on the `KafkaNodePool`(s) custom resource with `dual|kraft|zk` values to represent the the desired "direction" in the migration process by the user:
+  * `dual`: allows to move the migration process forward to reach the point of the "dual-write" with KRaft controllers deployed but with metadata written to both ZooKeeper and KRaft. Brokers are also connected to both ZooKeeper nodes and KRaft controllers.
+  * `kraft`: allows to move the migration process forward from the "dual-write" to the end of the migration with ZooKeeper out of the picture.
+  * `zk`: allows to move the migration process back, when still possible, to have brokers working again with ZooKeeper ensemble only.
```

Comment:
It is super weird to have one annotation with completely different semantics (values). I would pick different annotations if they are really needed on both types of resources.

Reply:
Fede raised the same point but as I said that was an open point I had to change this annotation name. I am fine to go with `run-kraft-migration` (as suggested by Fede) for now.